src/driver/amdxdna/ve2_hwctx.c

```diff
@@ -613,3 +613,147 @@ void ve2_hwctx_fini(struct amdxdna_ctx *hwctx)
+		break;
+	case AMDXDNA_FW_BUF_LOG:
+		if (attach) {
+			hs.log_addr_high = upper_32_bits((u64)paddr);
+			hs.log_addr_low = lower_32_bits((u64)paddr);
+			hs.log_buf_size = buf_sz;
+		}
+		ret = aie_partition_write_privileged_mem(aie_dev, CERT_HANDSHAKE_OFF(col) +
+							 offsetof(struct handshake, log_addr_high),
+							 sizeof(hs.log_addr_high) +
+							 sizeof(hs.log_addr_high) +
+							 sizeof(hs.log_buf_size),
+							 (void *)&hs.log_addr_high);
+		break;
+	default:
+		ret = -EOPNOTSUPP;
+		break;
+	}
+
+	return ret;
+}
+
+int ve2_hwctx_config(struct amdxdna_ctx *hwctx, u32 type, u64 mdata_hdl, void *buf, u32 size)
+{
+	struct amdxdna_dev *xdna = hwctx->client->xdna;
+	struct amdxdna_client *client = hwctx->client;
+	struct amdxdna_gem_obj *abo, *mdata_abo;
+	struct fw_buffer_metadata *mdata;
+	u32 prev_buf_sz;
+	u64 buf_paddr;
+	int ret = 0;
+	u32 buf_sz;
+
+	/* Update fw's handshake shared memory with debug/trace buffer details */
+	switch (type) {
+	case DRM_AMDXDNA_CTX_ASSIGN_DBG_BUF:
+		mdata_abo = amdxdna_gem_get_obj(client, mdata_hdl, AMDXDNA_BO_DEV);
+		if (!mdata_abo || !mdata_abo->mem.kva) {
+			XDNA_ERR(xdna, "Get metadata bo %lld failed for type %d", mdata_hdl, type);
+			return -EINVAL;
+		}
+
+		mdata = (struct fw_buffer_metadata *)(mdata_abo->mem.kva);
+		if (!mdata) {
+			XDNA_ERR(xdna, "No metadata defined for bo %lld type %d", mdata_hdl, type);
+			amdxdna_gem_put_obj(mdata_abo);
+			return -EINVAL;
+		}
+
+		abo = amdxdna_gem_get_obj(client, mdata->bo_handle, AMDXDNA_BO_DEV);
+		if (!abo || !abo->mem.kva) {
+			XDNA_ERR(xdna, "Get bo %lld failed for type %d", mdata->bo_handle, type);
+			amdxdna_gem_put_obj(mdata_abo);
+			return -EINVAL;
+		}
+
+		for (u32 col = 0; col < hwctx->num_col; col++) {
+			buf_sz = mdata->uc_info[col].size;
+			if (buf_sz == 0)
+				continue;
+			buf_paddr = abo->mem.dev_addr + prev_buf_sz;
+			ret = ve2_update_handshake_pkt(hwctx, mdata->buf_type, buf_paddr, buf_sz,
+						       col, true);
+			if (ret < 0) {
+				XDNA_ERR(xdna, "hwctx config req %d with flag %d failed, err %d",
+					 type, mdata->buf_type, ret);
+				amdxdna_gem_put_obj(abo);
+				amdxdna_gem_put_obj(mdata_abo);
+				return ret;
+			}
+			prev_buf_sz += buf_sz;
+		}
+		XDNA_DBG(xdna, "Attached %d BO %lld to %s, ret %d", mdata->buf_type,
+			 mdata->bo_handle, hwctx->name, ret);
+
+		amdxdna_gem_put_obj(abo);
+		amdxdna_gem_put_obj(mdata_abo);
+		break;
+	case DRM_AMDXDNA_CTX_REMOVE_DBG_BUF:
+		mdata_abo = amdxdna_gem_get_obj(client, mdata_hdl, AMDXDNA_BO_DEV);
+		if (!mdata_abo || !mdata_abo->mem.kva) {
+			XDNA_ERR(xdna, "Get metadata bo %lld failed for type %d", mdata_hdl, type);
+			return -EINVAL;
+		}
+
+		mdata = (struct fw_buffer_metadata *)(mdata_abo->mem.kva);
+		if (!mdata) {
+			XDNA_ERR(xdna, "No metadata defined for bo %lld type %d", mdata_hdl, type);
+			amdxdna_gem_put_obj(mdata_abo);
+			return -EINVAL;
+		}
+
+		for (u32 col = 0; col < hwctx->num_col; col++) {
```

Comment:
could you remove duplicated code for different types, such as: get mdata_abo, above for loop etc.